### PR TITLE
Feature/if then else

### PR DESCRIPTION
**Pull request by pawbu**

### CODE REVIEW

File: java-vtl-script/src/main/java/no/ssb/vtl/script/functions/IfFunctionExpression.java [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+                return conditionToExpression.get(conditionExpression).resolve(bindings);
+            }
+        }
+
+        return defaultExpression.resolve(bindings);
+    }
+
+    @Override
+    public Class getVTLType() {
+        return defaultExpression.getVTLType();
```
Comment (hadrienk) [inline java-vtl-script/src/main/java/no/ssb/vtl/script/functions/IfFunctionExpression.java]: You might want to cache this value.

File: java-vtl-parser/src/main/antlr4/no/ssb/vtl/parser/VTL.g4 [hunk 2]
```diff
@@ -169,6 +170,13 @@ ISNOTNULL : 'is not null' ;
 
 nvlFunction : 'nvl' LPAR expression COMMA expression RPAR ;
 
+ifThenElseExpression: 'if' ifBody ('elseif' ifBody)* ('else' ifBodyExpression) ;
+
+ifBody : ifBodyExpression 'then' ifBodyExpression ;
+
+ifBodyExpression : ifThenElseExpression {notifyErrorListeners("cannot embed if");}
```
Comment (hadrienk) [inline java-vtl-parser/src/main/antlr4/no/ssb/vtl/parser/VTL.g4]: this message is a bit hard to understand.

File: java-vtl-parser/src/test/java/no/ssb/vtl/parser/IfThenElseParserTest.java [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+    }
+
+    @Test
+    public void testIfThenElse() throws Exception {
+        parse("if true then 1 elseif true then 2 else 3");
+    }
+
+    @Test(expected = Exception.class)
+    public void testMissingThen() throws Exception {
+        parse("if true elseif true then 2 else 3");
```
Comment (hadrienk) [inline java-vtl-parser/src/test/java/no/ssb/vtl/parser/IfThenElseParserTest.java]: Does this work with `notifyErrorListeners("cannot embed if");` in VTL.g4?

Comment (hadrienk) [reply java-vtl-parser/src/test/java/no/ssb/vtl/parser/IfThenElseParserTest.java]: If so, maybe check on the specific exception?

File: java-vtl-script/src/main/java/no/ssb/vtl/script/functions/IfFunctionExpression.java [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+    private final VTLExpression defaultExpression;
+
+    private IfFunctionExpression(ImmutableMap<VTLExpression, VTLExpression> conditionToExpression,
+                                VTLExpression defaultExpression) {
+        this.conditionToExpression = conditionToExpression;
+        this.defaultExpression = defaultExpression;
+    }
+
+    @Override
+    public VTLObject resolve(Bindings bindings) {
```
Comment (hadrienk) [inline java-vtl-script/src/main/java/no/ssb/vtl/script/functions/IfFunctionExpression.java]: Super cool!

File: java-vtl-parser/src/main/antlr4/no/ssb/vtl/parser/VTL.g4 (hunk 2 above)
Comment (pawbu) [reply java-vtl-parser/src/main/antlr4/no/ssb/vtl/parser/VTL.g4]: OK, what do you propose?

File: java-vtl-parser/src/test/java/no/ssb/vtl/parser/IfThenElseParserTest.java (hunk 3 above)
Comment (pawbu) [reply java-vtl-parser/src/test/java/no/ssb/vtl/parser/IfThenElseParserTest.java]: The specific exception (`InputMismatchException`) is wrapped in `Exception` in `GrammarRule` class.

File: java-vtl-parser/src/main/antlr4/no/ssb/vtl/parser/VTL.g4 (hunk 2 above)
Comment (hadrienk) [reply java-vtl-parser/src/main/antlr4/no/ssb/vtl/parser/VTL.g4]: imbricated if-then-else expressions
or
value cannot be another if-then-else expression
?

File: java-vtl-parser/src/test/java/no/ssb/vtl/parser/IfThenElseParserTest.java (hunk 3 above)
Comment (hadrienk) [reply java-vtl-parser/src/test/java/no/ssb/vtl/parser/IfThenElseParserTest.java]: I see. I misunderstood. All good then.

**Comment by hadrienk** [comment]
I didn't catch this earlier, but this is not a function. Could you rename it to IfThenElseExpression and move it in a expression package?

File: java-vtl-parser/src/main/antlr4/no/ssb/vtl/parser/VTL.g4 (hunk 2 above)
Comment (pawbu) [reply java-vtl-parser/src/main/antlr4/no/ssb/vtl/parser/VTL.g4]: The latter sounds good :)

**Pull request merged**